### ISSUE TITLE
Divergent results between CPU and DirectML

### ISSUE THREAD

**Issue by vpenades** [microsoft/onnxruntime#9582]
**Describe the bug**

I've began leveraging OnnxRuntime, and started with a "super resolution" onnx model.

When I run on CPU, I get a correct result, but when I run on DirectML, I get a darker image than the one of the input. Everything is the same except enabling/disabling DirectML. 

**Urgency**

Not having consistent/reproducible results between providers certainly keeps me back from using onnxruntime.

**System information**
- OS Platform and Distribution: Win10
- ONNX Runtime version: 1.9
- Visual Studio version (if applicable): 2019
- GPU model and memory: nVidia GTX 1060 , 6GB

**To Reproduce**

Just setup a basic Image in, Image out session and run the model on CPU and DirectML.
I could not attach the model because it's heavy, I can point to a DL link if required.

**Expected behavior**
All providers should produce the same result based on the same input.

**Screenshots**
Input
![dog](https://user-images.githubusercontent.com/5433822/139149135-982b1575-1805-4c49-9b83-eb4e205903c3.jpeg)

CPU output:
![imagen](https://user-images.githubusercontent.com/5433822/139149216-8f5d823d-26ba-4a8b-bcff-20e0467dfbb5.png)

DirectML output:
![imagen](https://user-images.githubusercontent.com/5433822/139149276-c1de8853-1b8c-415a-8b73-963707e08459.png)

**Comment by vpenades** [comment]
@fdwr were you able to reproduce the problem?  just looking for a status report... no hurries.

**Comment by fdwr** [comment]
I may have grabbed the wrong .onnx model (given the lack of description in Netron). How large was your file, about 18MB's? The download size is crazy at 8GBs!

![image](https://user-images.githubusercontent.com/1809166/141930587-98456a2a-afff-45ff-91f2-1a2d176f037e.png)

**Comment by vpenades** [comment]
Yes, `realesrgan_256x256.onnx` at a size of `66'957'626 bytes` is the file I'm using.

I'm running it from C# , and I don't get any assert, neither in CPU nor in DirectML

**Comment by fdwr** [comment]
I'm not sure what's going on. My output image looks very blurry, for both CPU and DML :/ (good news is that neither looks any bright/dimmer than the other): 
![image](https://user-images.githubusercontent.com/1809166/142420558-ddfbde96-49c8-432a-ab9e-e3b5159f476e.png)

**Comment by vpenades** [comment]
that's reeeeeeally weird.

Just in case we're using exactly the same file, the SHA256 of my is realesrgan_256x256.onnx file is:

    86F3C4C530D0F739BA835D01E08343339CA2A18652A65E48649F53002A266717

Also, for inference `SessionOptions `on CPU, I'm using the default values, and for DML, I'm just calling `AppendExecutionProvider_DML` , no additional configuration.

Also, the input image is resized to match the 256x256 required size, and the RGB values normalized to floating values between 0 and 1

Other than that I think it's quite straightforward

**Comment by fdwr** [comment]
This is weird. If a run the graph with default optimizations, then I get a darker output. If I try to debug it though by finding where the first significant differences occur in the graph (by dumping out all the node tensors along the way to a directory full of .numpy tensor files and comparing them), then the difference no longer reproes 🤔, and both CPU and GPU look the same. So some transformation or fusion along the way is causing this.

![image](https://user-images.githubusercontent.com/1809166/142615393-3cd7f0b7-70bb-489f-b19c-202af423fb49.png)

**Comment by vpenades** [comment]
Interesting... so some optimizations are messing the model...

So, this is a model issue, or there's a bug in the DirectML optimized path?

**Comment by fdwr** [comment]
Seems so. It will be interesting to try ORT's GraphOptimizationLevel ORT_DISABLE_ALL to see if that makes a difference in the execution provider that calls DirectML...

**Comment by vpenades** [comment]
@fdwr I've tested my code with ORT_DISABLE_ALL and the result is good (same output as CPU)

I also tried ORT_ENABLE_BASIC which produces a dark image output.

**Comment by fdwr** [comment]
@vpenades: Thanks for that additional info.

**Comment by fdwr** [comment]
Just an update (underlying cause pending), if I do (a) or (b) or (c), it no longer reproes:

**(a)**

`{REG_INFO( 7, Conv, ... DmlGraphSupport::NotSupported)},​`
`​{REG_INFO( 11, Conv, ... DmlGraphSupport::NotSupported)},​`

**(b)**

In ExecutionProvider:

`​//#define ENABLE_GRAPH_COMPILATION​`

In GraphPartitioner.cpp

```
​PartitionGraph(...)
...
#if 0
for (const auto& input : partition->GetInputs())
{​
```

`(c)`

In conv_mul_fusion.cc

Return `false` from `​ConvMulFusion::SatisfyCondition()​`.

**Comment by vpenades** [comment]
@fdwr glad you're having some progress with the issue.

I only use OnnxRuntime from the C# side, so I'm not digging into de C++ code.

But, from what I undertand, when OnnxRuntime loads a model, it's able to optimize the graph if some conditions are met.

So, from the code you've written, it's either `​ConvMulFusion::SatisfyCondition()​` that sometimes returns true when it should not.... or the fusion itself is wrong.

**Comment by vpenades** [comment]
I've tried with OnnxRuntime 1.11 and the issue is still present.

When I use the workaround `GraphOptimizationLevel = Microsoft.ML.OnnxRuntime.GraphOptimizationLevel.ORT_DISABLE_ALL;` as suggested, the problem is resolved... 

But I am not sure if this workaround has been considered as the proper solution or if it remains a workaround until the underlaying issue has been resolved.

**Comment by fdwr** [comment]
@vpenades It's a work-around. The underlying issue remains -_-.

**Comment by vpenades** [comment]
I guess that it's not been resolved yet because it's non trivial....  and maybe it's not possible to be resolved without paying a performance cost?

**Comment by fdwr** [comment]
@vpenades: Sorry we didn't fix it for 1.11, but it's in 1.12 (in the master branch and [release 1.2.0 branch](https://github.com/microsoft/onnxruntime/tree/rel-1.2.0)).

>I guess it's not been resolved yet because it's non trivial.
Alas, anything with large models is non-trivial 😅. My teammate found the issue with a *different* simpler model, and it applied to this one too.
